docs/concepts/flink-architecture.md

```diff
@@ -129,4 +167,108 @@ two main benefits:
+  jobs that have tasks running on this worker will fail; in a similar way, if
+  some fatal error occurs on the Flink Master, it will affect all jobs running
+  in the cluster.
+
+* **Other considerations**: having a pre-existing cluster saves a considerable
+  amount of time applying for resources and starting TaskManagers. This is
+  important in scenarios where the execution time of jobs is very short and a
+  high startup time would negatively impact the end-to-end user experience — as
+  is the case with interactive analysis of short queries, where it is desirable
+  that jobs can quickly perform computations using existing resources.
+
+<div class="alert alert-info"> <strong>Note:</strong> Formerly, a Flink Session Cluster was also known as a Flink Cluster in <i>session mode</i>. </div>
+
+### Flink Job Cluster
+
+* **Cluster Lifecycle**: in a Flink Job Cluster, the available cluster manager
+  (like YARN or Kubernetes) is used to spin up a cluster for each submitted job
+  and this cluster is available to that job only. Here, the client first
+  requests resources from the cluster manager to start the Flink Master and
+  submits the job to the Dispatcher running inside this process. TaskManagers
+  are then lazily allocated based on the resource requirements of the job. Once
+  the job is finished, the Flink Job Cluster is torn down.
+
+* **Resource Isolation**: a fatal error in the Flink Master only ever affects
+  one job in a Flink Job Cluster.
+
+* **Other considerations**: because the ResourceManager has to apply and wait
+  for external resource management components to start the TaskManager
+  processes and allocate resources, Flink Job Clusters are more suited to large
+  jobs that are long-running, have high-stability requirements and are not
+  sensitive to higher startup times.
+
+<div class="alert alert-info"> <strong>Note:</strong> Formerly, a Flink Job Cluster was also known as a Flink Cluster in <i>job (or per-job) mode</i>. </div>
+
+### Flink Application Cluster
+
+* **Cluster Lifecycle**: a Flink Application Cluster is a dedicated Flink
+  cluster that only executes jobs from one Flink Application and where the
+  ``main()`` method runs on the cluster rather than the client. The job
+  submission is a one-step process: you don’t need to start a Flink cluster
+  first and then submit a job to the existing cluster session; instead, you
+  package your application logic and dependencies into a executable job JAR and
+  the cluster entrypoint ([ApplicationClusterEntryPoint]({{ site.baseurl
+  }}/api/java/index.html?org/apache/flink/container/entrypoint/StandaloneJobClusterEntryPoint.html))
+  is responsible for calling the ``main()`` method to extract the JobGraph.
+  This allows you to deploy a Flink Application like any other application on
+  Kubernetes, for example. The lifetime of a Flink Application Cluster is
+  therefore bound to the lifetime of the Flink Application.
```

Comment:
This section leaves me confused. What exactly is the difference between a Flink Job and a Flink Application?

Reply:
This is defined in the opening sentence of the "Application Execution" section. 

> A _Flink Application_ is any user program that spawns one or multiple Flink jobs from its ``main()`` method.

Do you think it would help to reiterate? In theory (AFAIU), an Application Cluster will also be bound to the lifetime of a job, since for now it doesn't support multi-execute applications; but in the future this may not be the case anymore.

Reply:
I think we have a terminology problem here. I've heard some folks using the term "application" to talk about the thing that evolves over time and as it gets deployed comes to life as a sequence of one flink job after another. I've also had it explained to me that "application cluster" is just updated terminology for "job cluster", and it means the same thing. And then this section is introducing yet another definition.

Reply:
There is definitely room for confusion. In this case, I was trying to reduce that by being consistent with the definition in the Glossary: https://ci.apache.org/projects/flink/flink-docs-master/concepts/glossary.html#flink-application

In the (near) future, it would probably be good to make this clearer by e.g. dropping the job cluster terminology and use just the new one — in the end:

"The "per job mode" is a bit of a (rarer) special case of the application, pulling out dataflow graph assembly, but otherwise is the same. To avoid overcomplicating this early, we were thinking to not have this in the concepts, but only in the detailed ops docs."

Users might be be too familiar with job clusters/per-job mode to just wipe it off sight and replace it, though.

Reply:
@alpinegizmo I think this confusion arose because people tried forcing "application mode" for the thing that was "job mode" before we had our terminology/features more sorted.

Doesn't the usage here more or less match your "thing that evolves over time and as it gets deployed comes to life as a sequence of one flink job after another"?

Reply:
@aljoscha Not at all. I was trying to evoke a definition of application that's akin to what one gets with a vvP deployment, whereas the definition here in this section is speaking about an application as a single program that spawns multiple jobs on a one-time basis.